src/GovUk.Education.ExploreEducationStatistics.Common/Services/ImportStatusService.cs

```diff
@@ -64,16 +88,85 @@ public async Task<bool> IsImportFinished(Guid releaseId, string dataFileName)
             return FinishedImportStatuses.Contains(importStatus.Status);
         }
 
-        private async Task<DatafileImport> GetImport(Guid releaseId, string dataFileName)
+        public async Task<bool> UpdateStatus(Guid releaseId, string origDataFileName, IStatus status)
         {
+            var import = await GetImport(releaseId, origDataFileName);
+            if (import.Status == IStatus.FAILED)
+            {
+                return false;
+            }
+
+            if (import.Status != status && import.Status != IStatus.COMPLETE)
+            {
+                import.PercentageComplete = status == IStatus.COMPLETE ? 100 : 0;
+                import.Status = status;
+
+                try
+                {
+                    await _table.ExecuteAsync(TableOperation.Replace(import));
+                }
+                catch (StorageException e)
+                {
+                    // If the table row has been updated in another thread subsequent
+                    // to being read above then an exception will be thrown - ignore and continue.
+                    // A similar approach will be required if optimistic locking is employed when & if
+                    // we switch from table storage to using db tables.
+                }
+            }
 
+            return true;
+        }
+
+        public async Task UpdateProgress(Guid releaseId, string origDataFileName, double percentageComplete)
+        {
+            var import = await GetImport(releaseId, origDataFileName);
+            if (import.PercentageComplete < (int) percentageComplete)
+            {
+                import.PercentageComplete = (int) percentageComplete;
```

Comment:
Would suggest clamping `percentageComplete` to 100 i.e. it can't exceed 100.

Reply:
Clamped ↕️